test/testfloat/testfloat.cpp

```diff
@@ -0,0 +1,423 @@
+        case Type::f32:
+        {
+            const FP value{value_.f32};
+            const FP expected{static_cast<uint32_t>(expected_bits)};
+            if (ignore_nans && std::isnan(expected.value))
+            {
+                if (expected.is_canonical_nan())
+                    return value.is_canonical_nan();
+                if (expected.is_arithmetic_nan())
+                    return value.is_arithmetic_nan();
+                return false;  // Signaling NaN should never happen.
+            }
+            return value == expected;
+        }
+
+        case Type::f64:
+        {
+            const FP value{value_.f64};
+            const FP expected{expected_bits};
+            if (ignore_nans && std::isnan(expected.value))
+            {
+                if (expected.is_canonical_nan())
+                    return value.is_canonical_nan();
+                if (expected.is_arithmetic_nan())
+                    return value.is_arithmetic_nan();
+                return false;  // Signaling NaN should never happen.
+            }
+            return value == expected;
+        }
+        }
+        __builtin_unreachable();
+    }
+
+    friend std::ostream& operator<<(std::ostream& os, const TypedValue& x)
+    {
+        os << std::hex << std::uppercase << std::setfill('0');
+        switch (x.type_)
+        {
+        case Type::i32:
+            return os << std::setw(8) << x.value_.as<uint32_t>();
+        case Type::i64:
+            return os << std::setw(16) << x.value_.i64;
+        case Type::f32:
+            return os << std::setw(8) << FP{x.value_.f32}.as_uint();
+        case Type::f64:
+            return os << std::setw(16) << FP{x.value_.f64}.as_uint();
+        }
+        __builtin_unreachable();
+    }
+};
+
+bool check(const FunctionDescription& func, fizzy::Instance& instance, const uint64_t inputs[],
+    bool ignore_nans)
+{
+    const auto report_failure = [&](auto result, auto expected) {
+        std::cerr << "FAILURE: " << result << " <-";
+        for (size_t i = 0; i < func.num_arguments; ++i)
+            std::cerr << ' ' << TypedValue{func.param_types[i], inputs[i]};
+        std::cerr << "\n         " << expected << " (expected)\n";
+    };
+
+    fizzy::Value args[2]{};
+    for (size_t i = 0; i < func.num_arguments; ++i)
+        args[i] = make_value(func.param_types[i], inputs[i]);
+
+    const auto r = fizzy::execute(
+        instance, func.idx, fizzy::span<const fizzy::Value>(args, func.num_arguments));
+
+    if (func.options == Options::TrapIsInvalidOperation)
+    {
+        const auto expected_exceptions = inputs[func.num_arguments + 1];
+        const bool invalid_operation = (expected_exceptions & InvalidOperation) != 0;
+
+        if (r.trapped == invalid_operation)
+            return true;
+
+        report_failure(r.trapped, invalid_operation);
+        return false;
+    }
+
+    const auto result = TypedValue{func.result_type, r.value};
```

Comment:
You don't check `r.trapped` in this case.
(also worth to check `r.has_value)`

Reply:
Good point.

Reply:
I added an assert because this is not testable.